### PR TITLE
gRPC conversions - ProvisionTokens

### PR DESCRIPTION
**Pull request by Joerger**
Add gRPC endpoints for Token CRUD methods.

Context passing for new endpoints refactored in #6118.

Merge after #6058, which this is based upon.

### CODE REVIEW

File: lib/auth/grpcserver.go [hunk 1]
```diff
@@ -1866,6 +1866,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
...
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	token, err := auth.ServerWithRoles.GetToken(ctx, req.Name)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	tokenV3, ok := token.(*types.ProvisionTokenV2)
+	if !ok {
+		return nil, trail.ToGRPC(trace.Errorf("encountered unexpected token type"))
```
Comment (awly) [inline lib/auth/grpcserver.go:1881]:
```suggestion
		return nil, trail.ToGRPC(trace.Errorf("encountered unexpected token type: %T", token))
```

File: lib/auth/grpcserver.go [hunk 2]
```diff
@@ -1866,6 +1866,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
...
+func (g *GRPCServer) GetToken(ctx context.Context, req *types.ResourceRequest) (*types.ProvisionTokenV2, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	token, err := auth.ServerWithRoles.GetToken(ctx, req.Name)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	tokenV3, ok := token.(*types.ProvisionTokenV2)
```
Comment (awly) [inline lib/auth/grpcserver.go:1879]: nit: `tokenV3`->`tokenV2`

File: lib/auth/grpcserver.go [hunk 3]
```diff
@@ -1866,6 +1866,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
...
+func (g *GRPCServer) UpsertToken(ctx context.Context, token *types.ProvisionTokenV2) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	if err = auth.ServerWithRoles.UpsertToken(ctx, token); err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	g.Debugf("%q token upserted", token.GetName())
```
Comment (awly) [inline lib/auth/grpcserver.go:1918]: don't put the actual token in the logs, it's a credential

**Review by awly — APPROVED** [review]

File: api/client/client.go [hunk 4]
```diff
@@ -997,3 +997,48 @@ func (c *Client) IsMFARequired(ctx context.Context, req *proto.IsMFARequiredRequ
 	}
 	return resp, nil
 }
+
+// GetToken returns a provision token by name.
+func (c *Client) GetToken(ctx context.Context, name string) (types.ProvisionToken, error) {
+	if name == "" {
+		return nil, trace.BadParameter("missing name")
```
Comment (fspmarshall) [inline api/client/client.go:1004]: Nit: Prefer error messages that provide a bit more context (e.g. `missing token name`, or even `cannot get token, missing name`).

**Review by fspmarshall — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

File: lib/auth/grpcserver.go [hunk 5]
```diff
@@ -1886,6 +1886,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
...
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	token, err := auth.ServerWithRoles.GetToken(ctx, req.Name)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	provisionTokenV2, ok := token.(*types.ProvisionTokenV2)
+	if !ok {
+		return nil, trail.ToGRPC(trace.Errorf("encountered unexpected token type: %T", token))
```
Comment (a-palchikov) [inline lib/auth/grpcserver.go:1901]: nit: I'd keep the error messages about mismatched types consistent with each other

File: lib/auth/grpcserver.go [hunk 6]
```diff
@@ -1886,6 +1886,75 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
...
+func (g *GRPCServer) UpsertToken(ctx context.Context, token *types.ProvisionTokenV2) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	if err = auth.ServerWithRoles.UpsertToken(ctx, token); err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+
+	g.Debugf("token upserted")
```
Comment (a-palchikov) [inline lib/auth/grpcserver.go:1938]: This is not logged by most other APIs - is this required? Same for `DeleteToken`

**Review by a-palchikov — APPROVED** [review]

**Pull request merged**